File: tests/llvm/hardlift/cpp/color_dodge_8.cc
#include <iostream>
#include <vector>
#include <chrono>

#include "random.h"

using namespace std;
using namespace std::chrono;

vector<vector<float>> a = random_matrix();
vector<vector<float>> b = random_matrix();

vector<vector<float>> color_dodge_8(vector<vector<float>> base, vector<vector<float>> active) {
    vector<vector<float>> out;
    int m = base.size();
    int n = base[0].size();
    for (int row = 0; row < m; row++) {
        vector<float> row_vec;
        for (int col = 0; col < n; col++) {
            float pixel;
            if (active[row][col] == 32)
                pixel = 32;
            else
                pixel = base[row][col] / (32 - active[row][col]);
            row_vec.push_back(pixel);
        }
        out.push_back(row_vec);
    }
    return out;
}


int main() {
    auto start_time = high_resolution_clock::now();
    color_dodge_8(a, b);
    auto end_time = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(end_time - start_time);

    cout << "Execution Time: " << duration.count() << " microseconds" << endl;
}

File: tests/llvm/tenspiler/benchmark/numba/color_dodge_8.py
from numba import njit


@njit
def color_dodge_8(base, active):
    output = []
    m = len(base)
    n = len(base[0])
    for i in range(m):
        curr_row = []
        for j in range(n):
            if active[i][j] == 255:
                pixel = 255
            else:
                pixel = base[i][j] / (255 - active[i][j])
            curr_row.append(pixel)
        output.append(curr_row)
    return output
